In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [6]:

# Relative path (since the file is in the same directory)
file_path = 'lending_data.csv'

# Load the CSV file into a DataFrame
df_lending_data = pd.read_csv(file_path)

# Display the first 10 rows
df_lending_data.head(10)

#/Users/acenyong/Desktop/Data_Science/CR/lending_data.csv

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   
5    10100.0          7.438            50600        0.407115                4   
6    10300.0          7.490            51100        0.412916                4   
7     8800.0          6.857            45100        0.334812                3   
8     9300.0          7.096            47400        0.367089                3   
9     9700.0          7.248            48800        0.385246                4   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0  
5                 1       20600            0  
6                 1       21100            0  
7                 0       15100            0  
8                 0       17400            0  
9                 0       18800            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [8]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending_data['loan_status']


# Separate the X variable, the features
X = df_lending_data.drop('loan_status', axis=1)

In [10]:
# Review the y variable Series
print(y.head()) 

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64


In [12]:
# Review the X variable DataFrame
print(X.head()) 

   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000  


### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [14]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [16]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
model = LogisticRegression(random_state=1)

# Fit the model using training data
model.fit(X_train, y_train)

# Now, the model is trained on the data
print("Model trained successfully!")

Model trained successfully!


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [18]:
# Make a prediction using the testing data

y_pred = model.predict(X_test)

# Print the predicted values for the first few entries
print(f"Predictions for the first few test samples: {y_pred[:10]}")

Predictions for the first few test samples: [0 0 0 0 0 0 0 0 0 0]


### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [24]:
# Import necessary modules for evaluation
from sklearn.metrics import confusion_matrix, classification_report

# Generate the confusion matrix
cm = confusion_matrix(y_test, y_pred)

# Print the confusion matrix
print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[14924    77]
 [   31   476]]


In [28]:
# Print the classification report for the model
cr = classification_report(y_test, y_pred)
print("\nClassification Report:")
print(cr)


Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     15001
           1       0.86      0.94      0.90       507

    accuracy                           0.99     15508
   macro avg       0.93      0.97      0.95     15508
weighted avg       0.99      0.99      0.99     15508



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 
1. Performance for Label 0 (healthy loan):

    Precision (1.00): Precision measures how many of the predicted healthy loans (0) are actually correct. A precision of 1.00 means that all of the predicted healthy loans were truly healthy loans.
    Recall (0.99): Recall measures how many of the actual healthy loans were correctly predicted. A recall of 0.99 means the model correctly identified 99% of all the actual healthy loans.
    F1-score (1.00): The F1-score is the harmonic mean of precision and recall. A score of 1.00 indicates perfect balance between precision and recall for healthy loans, meaning the model is very good at identifying this class.

2. Performance for Label 1 (high-risk loan):

    Precision (0.86): Precision for the high-risk loans is 0.86, meaning that 86% of the loans predicted as high-risk were actually high-risk. While this is good, it’s lower than for the healthy loan class.
    Recall (0.94): Recall for high-risk loans is 0.94, indicating that the model correctly identified 94% of the actual high-risk loans. This is quite high and suggests that the model is good at identifying most of the high-risk loans.
    F1-score (0.90): The F1-score for high-risk loans is 0.90, which indicates a solid balance between precision and recall, but there’s some room for improvement compared to the healthy loan class.

3. Overall Model Performance:

    Accuracy (0.99): The model has a very high overall accuracy, correctly predicting 99% of the total samples.
    Macro avg (0.93 Precision, 0.97 Recall, 0.95 F1-score): This is the average performance across both classes (healthy and high-risk loans). The model performs slightly better at identifying the healthy loans, but the macro averages indicate good overall performance.
    Weighted avg (0.99 Precision, 0.99 Recall, 0.99 F1-score): The weighted average takes the imbalance in the class distribution into account, and it shows that the model's overall performance, weighted by the number of samples in each class, is excellent.

---